code/game/objects/items/weapons/RCD.dm

```diff
@@ -10,6 +6,205 @@
+	var/delay = 0
+	/// The message (if any) to send the user when the action starts.
+	var/start_message
+	/// The effect (if any) to create when the action starts.
+	var/start_effect_type
+	/// The effect (if any) to create when the action completes.
+	var/end_effect_type
+	/// The mode the RCD must be in.
+	var/mode
+
+/// Attempt the action. This should not need to be overridden.
+/datum/rcd_act/proc/try_act(atom/A, obj/item/rcd/rcd, mob/user)
+	if(!can_act(A, rcd, user))
+		return FALSE
+	// We don't use the sound effect from use_tool because RCDs have a different sound effect for the start and end.
+	playsound(get_turf(rcd), 'sound/machines/click.ogg', 50, TRUE)
+	if(!rcd.tool_use_check(user, cost))
+		return FALSE
+	if(start_message)
+		to_chat(user, start_message)
+	var/start_effect
+	if(start_effect_type)
+		start_effect = new start_effect_type(get_turf(A))
+	if(!rcd.use_tool(A, user, delay, cost))
+		if(start_effect)
+			qdel(start_effect)
+		return FALSE
+	if(start_effect)
+		qdel(start_effect)
+	// If time elapsed, check our preconditions again.
+	if(delay && !can_act(A, rcd, user))
+		return FALSE
+	if(end_effect_type)
+		new end_effect_type(get_turf(A))
+	playsound(get_turf(rcd), 'sound/items/deconstruct.ogg', 50, TRUE)
+	act(A, rcd, user)
+	return TRUE
+
+/// Test to see if the act is possible. You should usually override this.
+/datum/rcd_act/proc/can_act(atom/A, obj/item/rcd/rcd, mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+	return rcd.mode == mode
+
+/// Perform the act. You should usually override this.
+/datum/rcd_act/proc/act(atom/A, obj/item/rcd/rcd, mob/user)
+	return
+
+/datum/rcd_act/place_floor
+	mode = MODE_TURF
+	cost = 1
+	start_message = "Building floor..."
+	end_effect_type = /obj/effect/temp_visual/rcd_effect/end
+
+/datum/rcd_act/place_floor/can_act(atom/A, obj/item/rcd/rcd, mob/user)
+	if(!..())
+		return FALSE
+	return isspaceturf(A) || istype(A, /obj/structure/lattice)
+
+/datum/rcd_act/place_floor/act(atom/A, obj/item/rcd/rcd, mob/user)
+	var/turf/AT = get_turf(A)
+	AT.ChangeTurf(/turf/simulated/floor/plating)
+
+/datum/rcd_act/place_wall
+	mode = MODE_TURF
+	cost = 3
+	start_message = "Building wall..."
+	delay = 2 SECONDS
+	start_effect_type = /obj/effect/temp_visual/rcd_effect/short
+	end_effect_type = /obj/effect/temp_visual/rcd_effect/end
+
+/datum/rcd_act/place_wall/can_act(atom/A, obj/item/rcd/rcd, mob/user)
+	if(!..())
+		return FALSE
+	return isfloorturf(A)
+
+/datum/rcd_act/place_wall/act(atom/A, obj/item/rcd/rcd, mob/user)
+	var/turf/AT = get_turf(A)
+	AT.ChangeTurf(/turf/simulated/wall)
+
+/datum/rcd_act/place_airlock
+	mode = MODE_AIRLOCK
+	cost = 10
+	start_message = "Building airlock..."
+	delay = 5 SECONDS
+	start_effect_type = /obj/effect/temp_visual/rcd_effect
+	end_effect_type = /obj/effect/temp_visual/rcd_effect/end
+
+/datum/rcd_act/place_airlock/can_act(atom/A, obj/item/rcd/rcd, mob/user)
+	if(!..())
+		return FALSE
+	return isfloorturf(A) && !(locate(/obj/machinery/door/airlock) in A.contents)
```

Comment:
```suggestion
	return isfloorturf(A) && !(locate(/obj/machinery/door/airlock) in A.contents)
```
I don't think we need locate for this. I think this is just for byond sugar reasons.